File: Day257_2025/Q257_Leetcode.py
from typing import List


class Solution:
    def spellchecker(self, wordlist: List[str], queries: List[str]) -> List[str]:
        def to_lower(word):
            return word.lower()

        def de_vowel(word):
            return ''.join('*' if c in 'aeiou' else c for c in word)

        exact = set(wordlist)
        case_map = {}
        vowel_map = {}

        for word in wordlist:
            lower = word.lower()
            devowel = de_vowel(lower)
            if lower not in case_map:
                case_map[lower] = word
            if devowel not in vowel_map:
                vowel_map[devowel] = word

        result = []
        for query in queries:
            if query in exact:
                result.append(query)
            else:
                lower = query.lower()
                devowel = de_vowel(lower)
                if lower in case_map:
                    result.append(case_map[lower])
                elif devowel in vowel_map:
                    result.append(vowel_map[devowel])
                else:
                    result.append("")

        return result

File: Day257_2025/Q257_Leetcode.java
import java.util.*;

class Q257_Leetcode {
    private String toLower(String s) {
        return s.toLowerCase();
    }

    private String deVowel(String s) {
        return s.replaceAll("[aeiou]", "*");
    }

    public String[] spellchecker(String[] wordlist, String[] queries) {
        Set<String> exact = new HashSet<>();
        Map<String, String> caseMap = new HashMap<>();
        Map<String, String> vowelMap = new HashMap<>();

        for (String word : wordlist) {
            exact.add(word);
            String lower = word.toLowerCase();
            String devowel = lower.replaceAll("[aeiou]", "*");

            caseMap.putIfAbsent(lower, word);
            vowelMap.putIfAbsent(devowel, word);
        }

        String[] result = new String[queries.length];

        for (int i = 0; i < queries.length; i++) {
            String query = queries[i];

            if (exact.contains(query)) {
                result[i] = query;
            } else {
                String lower = query.toLowerCase();
                String devowel = lower.replaceAll("[aeiou]", "*");

                if (caseMap.containsKey(lower)) {
                    result[i] = caseMap.get(lower);
                } else if (vowelMap.containsKey(devowel)) {
                    result[i] = vowelMap.get(devowel);
                } else {
                    result[i] = "";
                }
            }
        }

        return result;
    }
}
